pysmt/optimization/optimizer.py

```diff
@@ -151,49 +284,118 @@ def optimize(self, cost_function, strategy='ub',
 
         `step_size` the minimum reolution for finding a solution. The
         optimum will be found in the proximity of `step_size`
-
         """
+        rt = None, None
+        if goal.is_maximization_goal() or goal.is_minimization_goal():
+            rt = self._optimize(goal, strategy,
+                             feasible_solution_callback,
+                             step_size, **kwargs)
+        else:
+            raise GoalNotSupportedError("ExternalOptimizerMixin", goal)
+        return rt
+
+
+    def _optimize(self, goal, strategy,
+                 feasible_solution_callback,
+                 step_size, **kwargs):
+
+        current = OptSearchInterval(goal, self._comparation_functions)
+        first_step = True
+        while not current.empty():
+            if not first_step:
+                if strategy == "linear":
+                    lin_assertions = current.linear_search_cut(step_size)
+                elif strategy == "binary":
+                    lin_assertions = current.binary_search_cut()
+                else:
+                    raise PysmtValueError("Unknown optimization strategy '%s'" % strategy)
+                status = self.solve(assumptions=[lin_assertions])
+            else:
+                status = self.solve()
+                first_step = False
+            if status:
+                model = self.get_model()
+                current.search_is_sat(model)
+            else:
+                current.search_is_unsat()
+
+
+
+    def _optimize_max_min(self, goal, strategy,
```

Comment:
Is this still needed?

Reply:
no, but until I finish the implementation of the new classes I prefer not to remove the old ones